src/integer/mat_z/sort.rs

```diff
@@ -0,0 +1,296 @@
+// Mozilla Foundation. See <https://mozilla.org/en-US/MPL/2.0/>.
+
+//! Contains functions to sort [`MatZ`] by.
+
+use super::MatZ;
+use crate::{
+    error::MathError,
+    traits::{GetNumColumns, GetNumRows},
+    utils::sort::quicksort,
+};
+
+impl MatZ {
+    #[allow(clippy::needless_range_loop)]
+    /// Sorts the matrix based on some via `cond_func` specified condition of its columns.
+    ///
+    /// This condition is usually a norm with the described input-output behaviour.
+    ///
+    /// Paramters:
+    /// - `cond_func`: computes values implementing [`PartialOrd`] over the columns of the specified matrix.
+    /// These values are then used to re-order / sort the rows of the matrix.
+    ///
+    /// Returns an empty `Ok` if the action could be performed successfully.
+    /// A [`MathError`] is returned if the execution of `cond_func` returned an error.
+    ///
+    /// # Examples
+    /// ## Use a build-in function as condition
+    /// ```
+    /// use qfall_math::integer::MatZ;
+    /// use std::str::FromStr;
+    /// let mat = MatZ::from_str("[[3,2,1]]").unwrap();
+    /// let cmp = MatZ::from_str("[[1,2,3]]").unwrap();
+    ///
+    /// let sorted = mat.sort_by_column(MatZ::norm_eucl_sqrd).unwrap();
+    ///
+    /// assert_eq!(cmp, sorted);
+    /// ```
+    /// ## Use a custom function as condition
+    /// This function needs to take a column vector as input and output a type implementing [`PartialOrd`]
+    /// ```
+    /// use qfall_math::{integer::{MatZ, Z}, error::MathError, traits::{GetNumRows, GetEntry}};
+    /// use std::str::FromStr;
+    /// let mat = MatZ::from_str("[[3,2,1]]").unwrap();
+    /// let cmp = MatZ::from_str("[[1,2,3]]").unwrap();
+    ///
+    /// fn custom_cond_func(matrix: &MatZ) -> Result<Z, MathError> {
+    ///     let mut sum = Z::ZERO;
+    ///     for row in 0..matrix.get_num_rows() {
+    ///         sum = sum + matrix.get_entry(row, 0)?;
+    ///     }
+    ///     Ok(sum)
+    /// }
+    ///
+    /// let sorted = mat.sort_by_column(custom_cond_func).unwrap();
+    ///
+    /// assert_eq!(cmp, sorted);
+    /// ```
+    ///
+    /// # Errors and Failures
+    /// - Returns a [`MathError`] of the same type as `cond_func` if the execution of `cond_func` fails.
+    pub fn sort_by_column<T>(
+        &self,
+        cond_func: fn(&Self) -> Result<T, MathError>,
+    ) -> Result<Self, MathError>
+    where
+        T: PartialOrd,
+    {
+        let mut lengths = vec![];
+        for col in 0..self.get_num_columns() {
+            lengths.push(cond_func(&self.get_column(col).unwrap())?);
+        }
+        let length = lengths.len() - 1;
+        let mut id_vec: Vec<usize> = (0..self.get_num_columns() as usize).collect();
+        quicksort(&mut lengths, 0, length, &mut id_vec);
+
+        let mut out = Self::new(self.get_num_rows(), self.get_num_columns()).unwrap();
+        for col in 0..id_vec.len() {
+            out.set_column(col, self, id_vec[col]).unwrap();
+        }
```

Comment:
Rather do something like this than surpress the clippy warning. Also for the other case
```suggestion
        for (col, elem) in id_vec.iter().enumerate() {
            out.set_column(col, self, *elem).unwrap();
        }
```